### PR TITLE
[WIP] Remove various transmutes

### PR DESCRIPTION
**Pull request by nvzqz**
This PR removes various transmutes that can either be replaced with casts or safe code.

### CODE REVIEW

**Comment by petrochenkov** [comment]
I don't know what is the official policy and half of the changed cases looks better with `transmute` to me.
r? @bluss

File: src/librustc/ty/mod.rs [hunk 1]
```diff
@@ -592,7 +591,8 @@ impl<'tcx> serialize::UseSpecializedDecodable for &'tcx Slice<Ty<'tcx>> {}
 impl<T> Slice<T> {
     pub fn empty<'a>() -> &'a Slice<T> {
         unsafe {
-            mem::transmute(slice::from_raw_parts(0x1 as *const T, 0))
+            let slice = slice::from_raw_parts(0x1 as *const T, 0);
+            &*(slice as *const _ as *const _)
```
Comment (bluss) [inline src/librustc/ty/mod.rs]: This one in particular I don't know why transmute isn't the best way. We need some way to do the `&[X] -> &NewType([X])` cast.

**Comment by bluss** [comment]
Yes, we want to set a good example and not use transmute, though I agree many of them look better with transmute; with transmute, we also have some size checking that the pointer casts don't have.
But we don't have any pointee size check in either case.

So it's almost like we want better primitives that are transmute-like. Previous removal of them was in #14069 but there are frankly not many there that are worth bringing back, except transmute_lifetime, and whatever new ones we need.

File: src/librustc/ty/mod.rs (hunk 1 above)
Comment (nvzqz) [reply src/librustc/ty/mod.rs]: This still performs the same cast, just without transmuting the fat pointer. Can you explain further why transmute would be preferred here?

**Comment by shepmaster** [comment]
Happy new year from triage, @nvzqz! A reminder that this is awaiting your feedback!

**Comment by nvzqz** [comment]
@bluss I'm of the opinion that `mem::transmute` should be reserved for certain cases, such as ones that warrant a size check. The nomicon states that it's very easy to cause UB via transmutes, and I think that the Rust standard library/ies should be somewhat of a role model in how it uses them.

I agree that some of these _look_ better as transmutes but unless there's necessary size checks, alternatives should be considered.

@shepmaster Happy new year ❤️

**Comment by bluss** [comment]
@nvzqz Sure, I'm also partial to that. I don't really dislike these changes, but I hope we can find something even better for Rust than these myriads of double `as` casts.

File: src/libstd/sys/unix/ext/net.rs [hunk 2]
```diff
@@ -213,7 +213,9 @@ impl SocketAddr {
 
     fn address<'a>(&'a self) -> AddressKind<'a> {
         let len = self.len as usize - sun_path_offset();
-        let path = unsafe { mem::transmute::<&[libc::c_char], &[u8]>(&self.addr.sun_path) };
+        let path: &[libc::c_char] = unsafe {
+            &*(&self.addr.sun_path as *const _ as *const _)
```
Comment (bluss) [inline src/libstd/sys/unix/ext/net.rs]: This one is actually wrong — the original is tricky. `self.addr.sun_path` is an array and the transmute with explicit type is using the coercion from array to slice. This doesn't happen when we cast the reference to array in the new code.

I think the elided types were close to obscuring this bug.

Reinterpreting `&[u8] <-> &[c_char]` seems like it should be a function.

Comment (bluss) [reply src/libstd/sys/unix/ext/net.rs]: And the old target type was `&[u8]`, in this code it is instead `&[c_char]`. They should switch places. Convert from c_char to u8.

Comment (nvzqz) [reply src/libstd/sys/unix/ext/net.rs]: I see what you mean. Thanks a lot @bluss, you're awesome 👍🏻. Thanks for everything you do for this project. I'm make the appropriate changes tomorrow.

**Comment by shepmaster** [comment]
Heads up from triage, @bluss! Looks like there are new commits to be reviewed!

File: src/bootstrap/cache.rs [hunk 3]
```diff
@@ -114,43 +113,49 @@ impl Deref for Interned<String> {
     type Target = str;
     fn deref(&self) -> &'static str {
         let l = INTERNER.strs.lock().unwrap();
-        unsafe { mem::transmute::<&str, &'static str>(l.get(*self)) }
+        let s: &str = l.get(*self).as_ref();
```
Comment (eddyb) [inline src/bootstrap/cache.rs]: What's `as_ref` called on? If it's `&String` then you should remove `.as_ref()` and let it coerce.

Comment (eddyb) [reply src/bootstrap/cache.rs]: Same for all the other 5 cases in this file.

File: src/librustc/ty/mod.rs [hunk 4]
```diff
@@ -592,7 +591,8 @@ impl<'tcx> serialize::UseSpecializedDecodable for &'tcx Slice<Ty<'tcx>> {}
 impl<T> Slice<T> {
     pub fn empty<'a>() -> &'a Slice<T> {
         unsafe {
-            mem::transmute(slice::from_raw_parts(0x1 as *const T, 0))
+            let slice = slice::from_raw_parts(0x1 as *const T, 0);
+            &*(slice as *const [T] as *const Slice<T>)
```
Comment (eddyb) [inline src/librustc/ty/mod.rs]: This should be `&*(mem::align_of::<T>() as *const [T; 0] as *const [T] as *const Slice<T>)`.

**Comment by shepmaster** [comment]
Ping from triage @nvzqz ! We haven't heard from you in over a week; do you think you'll be able to respond to the feedback soon?

**Comment by shepmaster** [comment]
Thanks for the PR, @nvzqz ! Since we haven't heard from you in a few weeks, I'm going to go ahead and close this PR for now. Please feel free to reopen when you are able to address the last bit of feedback.

**Pull request closed without merging**